File: math/find-the-closest-palindrome.cpp
/*
  Given a string n representing an integer, return the closest integer (not including itself), which is a palindrome. If there is a tie, return the smaller one.
  
  The closest is defined as the absolute difference minimized between two integers.
  
  Example 1:
    Input: n = "123"
    Output: "121"
  
  Example 2:
    Input: n = "1"
    Output: "0"
    Explanation: 0 and 2 are the closest palindromes but we return the smallest which is 0.
*/
class Solution {
public:
    string nearestPalindromic(string n) {
        long long original = stoll(n);
        int length = n.length();
        
        vector<long long> candidates;
        
        string half = n.substr(0, (length + 1) / 2);
        long long halfNum = stoll(half);
        
        for (int i = -1; i <= 1; ++i) {
            candidates.push_back(create_palindrome(to_string(halfNum + i), length));
        }
        
        candidates.push_back(static_cast<long long>(pow(10, length - 1)) - 1); 
        candidates.push_back(static_cast<long long>(pow(10, length)) + 1);    
        
        long long minDiff = LLONG_MAX;
        long long closest = LLONG_MAX;
        
        for (auto candidate : candidates) {
            if (candidate == original) continue;
            
            long long diff = abs(candidate - original);
            if (diff < minDiff || (diff == minDiff && candidate < closest)) {
                closest = candidate;
                minDiff = diff;
            }
        }
        
        return to_string(closest);
    }
private:
    long long create_palindrome(const string& half, int length) {
        string secondHalf = half;
        reverse(secondHalf.begin(), secondHalf.end());
    
        if (length % 2 == 0) return stoll(half + secondHalf);
        else return stoll(half + secondHalf.substr(1));
    }
};

File: math/find-the-closest-palindrome.cs
/*
  Given a string n representing an integer, return the closest integer (not including itself), which is a palindrome. If there is a tie, return the smaller one.
  
  The closest is defined as the absolute difference minimized between two integers.
  
  Example 1:
    Input: n = "123"
    Output: "121"
  
  Example 2:
    Input: n = "1"
    Output: "0"
    Explanation: 0 and 2 are the closest palindromes but we return the smallest which is 0.
*/
public class Solution {
    public string NearestPalindromic(string n) {
        long original = long.Parse(n);
        int len = n.Length;
        
        HashSet<long> candidates = [];
        
        string half = n[0..((len + 1) / 2)];
        long halfNum = long.Parse(half);
        
        for (int i = -1; i <= 1; ++i) {
            candidates.Add(CreatePalindrome((halfNum + i).ToString(), len));
        }
        
        candidates.Add((long)Math.Pow(10, len - 1) - 1);
        candidates.Add((long)Math.Pow(10, len) + 1);
        candidates.Remove(original);

        long closest = long.MaxValue;
        long minDiff = long.MaxValue;
        
        foreach (long candidate in candidates) {
            long diff = Math.Abs(candidate - original);
            if (diff < minDiff || (diff == minDiff && candidate < closest)) {
                closest = candidate;
                minDiff = diff;
            }
        }
        
        return closest.ToString();
    }

    private static long CreatePalindrome(string half, int len) {
        Span<char> halfArr = half.ToCharArray();
        halfArr.Reverse();

        string secondHalf = new string(halfArr);
        return long.Parse(half + secondHalf[(len % 2)..]);
    }
}
